scheduler-bindings/src/lib.rs

```diff
@@ -0,0 +1,210 @@
+//! and is done via `rts-alloc` crate. It is also possible the external
+//! pack process allocates memory to pass to agave, BUT it will still be
+//! the responsibility of the external pack process to free that memory.
+
+/// Reference to a transaction that can shared safely across processes.
+#[repr(C)]
+pub struct SharableTransaction {
+    /// Offset within the shared memory allocator.
+    pub offset: usize,
+    /// Length of the transaction in bytes.
+    pub length: u32,
+}
+
+/// Reference to an array of Pubkeys that can be shared safely across processes.
+#[repr(C)]
+pub struct SharablePubkeys {
+    /// Offset within the shared memory allocator.
+    pub offset: usize,
+    /// Number of pubkeys in the array.
+    /// IF 0, indicates no pubkeys and no allocation needing to be freed.
+    pub num_pubkeys: u32,
+}
+
+/// Message: [TPU -> Pack]
+/// TPU passes transactions to the external pack process.
+/// This is also a transfer of ownership of the transaction:
+///   the external pack process is responsible for freeing the memory.
+pub type TpuToPackMessage = SharableTransaction;
+
+/// Message: [Agave -> Pack]
+/// Agave passes leader status to the external pack process.
+#[repr(C)]
+pub struct ProgressMessage {
+    /// The slot the status is for.
+    pub slot: u64,
+    /// The current progress of the slot in percentage.
+    /// Negative values indicate approximate time until the first leader slot
+    /// begins.
+    pub progress: i16,
+    /// The number of compute units packed in the current block, if leader.
+    pub total_compute_units: u64,
+}
+
+/// The maximum number of transactions that can be included in a single
+/// [`PackToWorkerMessage`].
+pub const MAX_TRANSACTIONS_PER_PACK_MESSAGE: usize = 32;
+
+/// Special value for [`PackToWorkerMessage::slot`] indicating the transactions
+/// can be processed in any slot in agave.
+pub const ANY_SLOT: u64 = u64::MAX;
+
+/// Message: [Pack -> Worker]
+/// External pack processe passes transactions to worker threads within agave.
+///
+/// These messages do not transfer ownership of the transactions.
+/// The external pack process is still responsible for freeing the memory.
+#[repr(C)]
+pub struct PackToWorkerMessage {
+    /// Flags on how to handle this message.
+    /// See [`pack_message_flags`] for details.
+    pub flags: u16,
+    /// If [`pack_message_flags::RESOLVE`] flag is not set, this is the slot
+    /// the transactions can be processed in. If the flag is set to
+    /// [`ANY_SLOT`], the transactions can be processed in any slot in agave.
+    pub slot: u64,
+    /// The number of transactions in this message.
+    /// MUST be in the range [1, [`MAX_TRANSACTIONS_PER_PACK_MESSAGE`]].
+    pub num_transactions: u8,
+    /// Transactions in the message. Only the first `num_transactions`
+    /// entries are valid to read.
+    pub transactions: [SharableTransaction; MAX_TRANSACTIONS_PER_PACK_MESSAGE],
```

Comment:
Thought about making this an allocation from the external pack, which would allow us an arbitrary number of txs per message.
But I don't think there's enough value in that vs the static sized array to avoid additional indirection.

Reply:
We could start with variable length txs if unsure `MAX_TRANSACTIONS_PER_PACK_MESSAGE` will change often in the future.